src/irmin-pack/unix/gc.ml

```diff
@@ -0,0 +1,280 @@
+  }
+
+  let v ~root ~generation ~unlink ~offset ~dispatcher ~fm ~contents ~node
+      ~commit commit_key =
+    let unlink_result_file () =
+      let result_file = Irmin_pack.Layout.V3.gc_result ~root ~generation in
+      match Io.unlink result_file with
+      | Ok () -> ()
+      | Error (`Sys_error msg as err) ->
+          if msg <> Fmt.str "%s: No such file or directory" result_file then
+            [%log.warn
+              "Unlinking temporary files from previous failed gc. Failed with \
+               error %a"
+              Errs.pp err]
+    in
+    (* Unlink next gc's result file, in case it is on disk, for instance
+       after a failed gc. *)
+    unlink_result_file ();
+    (* function to track durations *)
+    let elapsed =
+      let c0 = Mtime_clock.counter () in
+      fun () -> Mtime_clock.count c0 |> Mtime.Span.to_s
+    in
+    (* internal promise for gc *)
+    let promise, resolver = Lwt.wait () in
+    (* start worker task *)
+    let task =
+      Async.async (fun () ->
+          let (_ : int63) =
+            Worker.run_and_output_result root commit_key ~generation
+          in
+          ())
+    in
+    {
+      root;
+      generation;
+      unlink;
+      offset;
+      task;
+      promise;
+      resolver;
+      elapsed;
+      dispatcher;
+      fm;
+      contents;
+      node;
+      commit;
+    }
+
+  let open_new_suffix ~end_offset { root; generation; _ } =
+    let open Result_syntax in
+    let path = Irmin_pack.Layout.V3.suffix ~root ~generation in
+    (* As the new suffix is necessarily in V3, the dead_header_size is
+       0. *)
+    let dead_header_size = 0 in
+    let auto_flush_threshold = 1_000_000 in
+    let auto_flush_callback x = Ao.flush x |> Errs.raise_if_error in
+    let* suffix =
+      Ao.open_rw ~path ~end_offset ~dead_header_size ~auto_flush_callback
+        ~auto_flush_threshold
+    in
+    Ok suffix
+
+  let transfer_latest_newies ~right_start_offset ~copy_end_offset t =
+    [%log.debug "Gc in main: transfer latest newies"];
+    let open Result_syntax in
+    let open Int63.Syntax in
+    let old_end_offset = Dispatcher.end_offset t.dispatcher in
+    let remaining = old_end_offset - copy_end_offset in
+    (* When opening the suffix in append_only we need to provide a
+       (real) suffix offset, computed from the global ones. *)
+    let suffix_end_offset = copy_end_offset - right_start_offset in
+    let* new_suffix = open_new_suffix ~end_offset:suffix_end_offset t in
+    Errors.finalise (fun _ ->
+        Ao.close new_suffix
+        |> Errs.log_if_error "GC: Close suffix after copy latest newies")
+    @@ fun () ->
+    let buffer = Bytes.create 8192 in
+    let read_exn = Dispatcher.read_exn t.dispatcher in
+    let append_exn = Ao.append_exn new_suffix in
+    let flush_and_raise () = Ao.flush new_suffix |> Errs.raise_if_error in
+    let* () =
+      Errs.catch (fun () ->
+          Worker.transfer_append_exn ~read_exn ~append_exn ~off:copy_end_offset
```

Comment:
I think this might be a code smell for how we are reusing worker code but no longer in the context of the worker. I kept it as-is for now, but we might want to revisit in the future.

Reply:
I think that the answer to this concern and @icristescu's about the duplication of `Args` is to create a third file (`gc_utils`?) that would contain these